### ISSUE TITLE
invalid type mismatch

### ISSUE THREAD

**Issue by wanderer** [WebAssembly/spec#290]
The following code generates this error when ran
`test.wast:21.28-21.42: type mismatch: expression has type i32 but the context requires i64`
code:
```
(module
  (func $test
    (param $a i64)
    (local $b i64)
    (param $c i64)
    (local $d i64)
    (param $e i64)
    (local $f i64)
    (param $g i64)
    (local $h i64)

    (param $i i64)
    (local $j i64)
    (param $k i64)
    (local $l i64)
    (param $m i64)
    (local $n i64)
    (param $o i64)

    (param $memIndex i32)
    (set_local $i (i64.and (get_local $i) (i64.const 4294967295)))
  )
)
```

if I remove one (and it doesn't matter which one) `param` or `local` it runs without error

**Comment by wanderer** [comment]
I think this is related too, to following test fails to return the correct value

```
(module
  (func $sub
    (param $a1 i64)
    (param $b1 i64)
    (param $c1 i64)
    (param $d1 i64)
    (local $carry i32)
    (param $i i32)
    (result i32)
    (get_local $i)
  )

  (export "sub" $sub)
)

(assert_return (invoke "sub"  (i64.const -1) (i64.const -1) (i64.const -1) (i64.const -1) (i32.const 24)) (i32.const 24))
```

**Comment by wanderer** [comment]
so when I arrange all the params to be at the top then the locals everything works

**Comment by rossberg-chromium** [comment]
Thanks. I introduced this bug when I removed the original index space
distinction between parameters and locals. They do not get reordered
properly. I will look into a fix.

**Comment by rossberg-chromium** [comment]
Addressed in #291 by making a misplaced `local` a syntax error.
